server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java

```diff
@@ -20,255 +20,137 @@
-
-        final RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, clusterState, null, null, 0L);
-        assertSame(Decision.NO, allocationDeciders.canAllocate(shardRouting, routingNode, allocation));
-        assertSame(Decision.NO, allocationDeciders.canRebalance(shardRouting, allocation));
-        assertSame(Decision.NO, allocationDeciders.canRemain(shardRouting, routingNode, allocation));
-        assertSame(Decision.NO, allocationDeciders.canAllocate(shardRouting, allocation));
-        assertSame(Decision.NO, allocationDeciders.canAllocate(indexMetadata, routingNode, allocation));
-        assertSame(Decision.NO, allocationDeciders.shouldAutoExpandToNode(indexMetadata, null, allocation));
-        assertSame(Decision.NO, allocationDeciders.canRebalance(allocation));
-        assertSame(Decision.NO, allocationDeciders.canForceAllocatePrimary(shardRouting, routingNode, allocation));
-
-        // debug decision should contain both individual decisions in a multi-decision
-        allocation.debugDecision(true);
-        final Decision expectedDebugDecision = new Decision.Multi().add(decisionOne).add(decisionTwo);
-        assertEquals(expectedDebugDecision, allocationDeciders.canAllocate(shardRouting, routingNode, allocation));
-        assertEquals(expectedDebugDecision, allocationDeciders.canRebalance(shardRouting, allocation));
-        assertEquals(expectedDebugDecision, allocationDeciders.canRemain(shardRouting, routingNode, allocation));
-        assertEquals(expectedDebugDecision, allocationDeciders.canAllocate(shardRouting, allocation));
-        assertEquals(expectedDebugDecision, allocationDeciders.canAllocate(indexMetadata, routingNode, allocation));
-        assertEquals(expectedDebugDecision, allocationDeciders.shouldAutoExpandToNode(indexMetadata, null, allocation));
-        assertEquals(expectedDebugDecision, allocationDeciders.canRebalance(allocation));
-        assertEquals(expectedDebugDecision, allocationDeciders.canForceAllocatePrimary(shardRouting, routingNode, allocation));
+        ShardRouting shardRouting = createShardRouting(index.getIndex());
+        RoutingNode routingNode = RoutingNodesHelper.routingNode("node", null);
+        DiscoveryNode discoveryNode = new DiscoveryNode("node", new TransportAddress(TransportAddress.META_ADDRESS, 0), Version.CURRENT);
+
+        List.<BiFunction<RoutingAllocation, AllocationDeciders, Decision>>of(
+            (allocation, deciders) -> deciders.canAllocate(shardRouting, allocation),
+            (allocation, deciders) -> deciders.canAllocate(shardRouting, routingNode, allocation),
+            (allocation, deciders) -> deciders.canAllocate(index, routingNode, allocation),
+            (allocation, deciders) -> deciders.canRebalance(allocation),
+            (allocation, deciders) -> deciders.canRebalance(shardRouting, allocation),
+            // (allocation, deciders) -> deciders.canRemain(shardRouting, routingNode, allocation),
```

Comment:
This is temporarily commented as `canRemain` has a somewhat different behavior: https://github.com/elastic/elasticsearch/blob/bb3addbd706c84e435bf66699544a2d94f810fda/server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDeciders.java#L91-L126

Reply:
ok, as long as we don't leave it like this too long :)